src/Bicep.Core/Emit/ExpressionConverter.cs

```diff
@@ -158,25 +158,25 @@ public LanguageExpression ConvertExpression(SyntaxBase expression)
 
         private LanguageExpression GetResourceNameExpression(ResourceDeclarationSyntax resourceSyntax)
         {
-            if (!(resourceSyntax.Body is ObjectSyntax objectSyntax))
+            if (resourceSyntax.ResolvedBody is not ObjectSyntax objectSyntax)
             {
                 // this condition should have already been validated by the type checker
-                throw new ArgumentException($"Expected resource syntax to have type {typeof(ObjectSyntax)}, but found {resourceSyntax.Body.GetType()}");
+                throw new ArgumentException($"Expected resource syntax to have type {typeof(ObjectSyntax)} or {typeof(IfExpressionSyntax)}, but found {resourceSyntax.ResolvedBody.GetType()}");
             }
-            
+
             // this condition should have already been validated by the type checker
             var namePropertySyntax = objectSyntax.SafeGetPropertyByName("name") ?? throw new ArgumentException($"Expected resource syntax body to contain property 'name'");
             return  ConvertExpression(namePropertySyntax.Value);
         }
 
         private LanguageExpression GetModuleNameExpression(ModuleDeclarationSyntax moduleSyntax)
         {
-            if (!(moduleSyntax.Body is ObjectSyntax objectSyntax))
+            if (moduleSyntax.ResolvedBody is not ObjectSyntax objectSyntax)
```

Comment:
Is objectSyntax necessary now that we use C# 9?

Reply:
Yeah - `(x is not Y y)` is a syntax introduced in C# 9 - to give yourself a non-null reference to an object `Y` outside of the if block. The `objectSyntax` reference here is being used on line 181.